<div style="border:solid Chocolate 2px; padding: 40px">
 
<b> Санат, привет!👋</b> 

Меня зовут Кирилл Васильев, я буду ревьюером твоего проекта. Я предлагаю общаться на «ты», но если привычнее на «вы», просто скажи об этом! 

Я буду оставлять комментарии в твоем проекте. Пожалуйста, не удаляй их, они могут быть полезны в случае повторной проверки.

Свои комментарии я буду обозначать <font color='green'>зеленым</font>, <font color='gold'>желтым</font> и <font color='red'>красным</font> цветами, например:
  
<br/> 

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Все отлично!👍:</b> В случае, если решение на отдельном шаге является полностью правильным. Здесь же я могу давать советы и предложения.
</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> В случае, когда решение на отдельном шаге станет еще лучше, если внести небольшие коррективы.
</div> 
 

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
 
   
<b>На доработку🤔:</b>
 В случае, когда решение на отдельном шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.
</div>  

<br/>
<div class="alert alert-info">
Рекомендую для твоих комментариев использовать синий или любой другой цветной фон. Я постараюсь ничего не пропустить, но если комментарии будут выделены, то это мне очень поможет!
</div>
    
Увидев у тебя неточность, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой руководитель будет поступать также. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.



# Анализ кол-ва выпускаемых игр разных жанров с 2000 по 2013

- Автор:Кенжебеков Санат da_123
- Дата:09.07.25

      
 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
<b>Некоторые замечания и рекомендации💡:</b> 

Предлагаю действительно написать заголовок проекта вместо `Напишите заголовок проекта здесь`
 </div>   

    

<div style="background-color: #e6f3ff; padding: 10px; border-radius: 5px;">
Добавил заголовок
</div>

### Цели и задачи проекта

<font color='#777778'>Перечислите цель проекта и задачи, которые вы решаете. Для этого можно использовать описание проекта, но полезнее будет сформулировать основную цель проекта самостоятельно.
    
В этой тетрадке используем датасет `/datasets/new_games.csv`, который содержит информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр. 
    Познакомимся с данными, сделаем вывод о полученных данных: данные какого объёма нам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных. Отметим другие особенности данных, которые мы обнаружим и на которые стоит обратить внимание при предобработке.
    Проведем фильтрацию данных в период с 2000 по 2013 год включительно, т.к. именно за этот период коллеги хотят изучить историю продаж игр. 
    Проведем категоризацию данных:
    - Разделим все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).
    - Разделим все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).
    - Выделим топ-7 платформ по количеству игр, выпущенных за весь требуемый период.
</font>

### Описание данных

<font color='#777778'>Данные /datasets/new_games.csv содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:
- Name — название игры.
- Platform — название платформы.
- Year of Release — год выпуска игры.
- Genre — жанр игры.
- NA sales — продажи в Северной Америке (в миллионах проданных копий).
- EU sales — продажи в Европе (в миллионах проданных копий).
- JP sales — продажи в Японии (в миллионах проданных копий).
- Other sales — продажи в других странах (в миллионах проданных копий).
- Critic Score — оценка критиков (от 0 до 100).
- User Score — оценка пользователей (от 0 до 10).
- Rating — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.</font>

### Содержимое проекта

<font color='#777778'>
- 1. Загрузка и знакомство с данными
- 2. Поиск пропущенных значений
- 3. Корректировка типа данных
- 4. Фильтрация данных
- 5. Категоризация данных
</font>

---

<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Отлично по описанию. Цели и задачи ясны. Контекст проекта будет понятен любому потенциальному читателю.
 
Чтобы текст встроках не слипался, можно оформить все либо пунктами, либо сделать пробел между строками в редакторе markdown, тогда строки перенесутся.
    
Пункты можно сделать так:
    
* раз
* два
* три


## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [1]:
# Пустые ячейки после каждого задания — примерное пространство для работы.
# Вы можете свободно добавлять или удалять ячейки по своему усмотрению в зависимости от логики и объёма работы.
# Импортируем библиотеку pandas
import pandas as pd

In [2]:
# Выгружаем данные из датасета new_games.csv в датафрейм games
games = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

- Познакомьтесь с данными: выведите первые строки и результат метода `info()`.


In [3]:
# Выводим информацию о датафрейме
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


In [4]:
games.head(20)

Name Platform  Year of Release  \
0                                     Wii Sports      Wii           2006.0   
1                              Super Mario Bros.      NES           1985.0   
2                                 Mario Kart Wii      Wii           2008.0   
3                              Wii Sports Resort      Wii           2009.0   
4                       Pokemon Red/Pokemon Blue       GB           1996.0   
5                                         Tetris       GB           1989.0   
6                          New Super Mario Bros.       DS           2006.0   
7                                       Wii Play      Wii           2006.0   
8                      New Super Mario Bros. Wii      Wii           2009.0   
9                                      Duck Hunt      NES           1984.0   
10                                    Nintendogs       DS           2005.0   
11                                 Mario Kart DS       DS           2005.0   
12                   Pokemon Gold/Pokemon Silver       GB           1999.0   
13                                       Wii Fit      Wii           2007.0   
14                            Kinect Adventures!     X360           2010.0   
15                                  Wii Fit Plus      Wii           2009.0   
16                            Grand Theft Auto V      PS3           2013.0   
17                 Grand Theft Auto: San Andreas      PS2           2004.0   
18                             Super Mario World     SNES           1990.0   
19  Brain Age: Train Your Brain in Minutes a Day       DS           2005.0   

           Genre  NA sales EU sales JP sales  Other sales  Critic Score  \
0         Sports     41.36    28.96     3.77         8.45          76.0   
1       Platform     29.08     3.58     6.81         0.77           NaN   
2         Racing     15.68    12.76     3.79         3.29          82.0   
3         Sports     15.61    10.93     3.28         2.95          80.0   
4   Role-Playing     11.27     8.89    10.22         1.00           NaN   
5         Puzzle     23.20     2.26     4.22         0.58           NaN   
6       Platform     11.28     9.14      6.5         2.88          89.0   
7           Misc     13.96     9.18     2.93         2.84          58.0   
8       Platform     14.44     6.94      4.7         2.24          87.0   
9        Shooter     26.93     0.63     0.28         0.47           NaN   
10    Simulation      9.05    10.95     1.93         2.74           NaN   
11        Racing      9.71     7.47     4.13         1.90          91.0   
12  Role-Playing      9.00     6.18      7.2         0.71           NaN   
13        Sports      8.92     8.03      3.6         2.15          80.0   
14          Misc     15.00     4.89     0.24         1.69          61.0   
15        Sports      9.01     8.49     2.53         1.77          80.0   
16        Action      7.02     9.09     0.98         3.96          97.0   
17        Action      9.43      0.4     0.41        10.57          95.0   
18      Platform     12.78     3.75     3.54         0.55           NaN   
19          Misc      4.74      9.2     4.16         2.04          77.0   

   User Score Rating  
0           8      E  
1         NaN    NaN  
2         8.3      E  
3           8      E  
4         NaN    NaN  
5         NaN    NaN  
6         8.5      E  
7         6.6      E  
8         8.4      E  
9         NaN    NaN  
10        NaN    NaN  
11        8.6      E  
12        NaN    NaN  
13        7.7      E  
14        6.3      E  
15        7.4      E  
16        8.2      M  
17          9      M  
18        NaN    NaN  
19        7.9      E

- Сделайте вывод о полученных данных: данные какого объёма вам предоставили, соответствуют ли они описанию, встречаются ли в них пропуски, используются ли верные типы данных.
- Отметьте другие особенности данных, которые вы обнаружили и на которые стоит обратить внимание при предобработке. Например, вы можете проверить названия столбцов: все ли названия отражают содержимое данных и прописаны в удобном для работы виде.
Датасет new_games.csv содержит 11 столбцов и 16956 строк.

Изучим типы данных и их корректность:

- **Числовые значения с плавающей запятой (float64).** 4 столбца имеют данный тип данных:
NA sales, Other sales содержат количество продаж в миллионах представлен типом float64. Это верное решение, так как они могут включать дробные значения. Year of Release представлен типом float64, что неправильно, так как год не может быть дробным значением. Это столбец логично поменять на тип int64 или меньшей разрядностью, но в нем встречаются пропуски и пока этого не получится. Может после предобработки пропусков мы к этому вернемся.
Critic Score тоже представлен типом float64, что отражает оценку критиков от 0 до 100. Открыв первые 20 строк, встречаются в основном целые значения с 0 после запятой. Логично поменять тип данных на int64 или меньшей разрядностью для оптимизации, но в этом столбце встречаются пропуски, поэтому этого сделать не получится.
- **Строковые данные (object).** Восемь столбцов имеют тип данных object:
Name, Platform, Genre, Rating содержат строковую информацию (название игры, название платформы, жанр игры и категории рейтинга ESRB), что логично для текстовых данных. Здесь тип данных object подходит. Однако для столбцов Platform, Genre можно использовать тип category, чтобы улучшить производительноть и оптимизировать память.
User Score представлен типом object, что неправильно, т.к. он отражает оценку пользователей от 0 до 10. Открыв первые 20 строк датафрейма, я вижу что имеются дробные значения, соответственно поменяем тип данных на float64.
EU sales, JP sales представлены типом object, что неправильно. Т.к. это количество продаж в миллионах, то нужно поменять тип на float64, как это представлено в NA sales.

После анализа типов данных видно, что значения во многих столбцах представлены некорректным типом данных. Это будем исправлять. 

На основе вывода метода .info() можно сделать следующие выводы о полноте данных:

Столбцы с пропусками (неполные данные):

- Name
- Year of Release
- Genre
- Critic Score
- User Score
- Rating

Все остальные 5 столбцов имеют 16956 заполненных строк

  
<div class="alert alert-success">
<h2> Комментарий ревьюера  <a class="tocSkip"> </h2>

👍 Все так. В данных есть пропуски, можно поправить по стилю названий колонок, а также обязательно нужно разобраться в проблемах с несоответствием типов данных содержанию столбцов.





  <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

`После анализа типов данных видно, что множество столбцов некорректно представлены.` - здесь слишком широкая формулировка, не ясно, что имеется ввиду. Почему некорректно, в чем ошибка?

<div style="background-color: #e6f3ff; padding: 10px; border-radius: 5px;">
Исправил
</div>

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [5]:
games.columns = games.columns.str.lower().str.replace(' ', '_')

print("\nНовые названия столбцов:", games.columns.tolist())
games.head()


Новые названия столбцов: ['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating']


name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  eu_sales jp_sales  other_sales  critic_score user_score rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Верно и оптимально по переименованию.  Отлично, что делаешь проверку результата преобразования.



### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [6]:
# Преобразование столбцов в категориальный тип
games['platform'] = games['platform'].astype('category')
games['genre'] = games['genre'].astype('category')

# Проверка типов данных
print(games[['platform', 'genre']].dtypes)

platform    category
genre       category
dtype: object


In [7]:
# Преобразование столбцов к типу float64
games['eu_sales'] = pd.to_numeric(games['eu_sales'], errors='coerce')
games['jp_sales'] = pd.to_numeric(games['jp_sales'], errors='coerce')
games['user_score'] = pd.to_numeric(games['user_score'], errors='coerce')

# Проверка типов данных
print(games[['user_score', 'eu_sales', 'jp_sales']].dtypes)
print(games['user_score'].unique())  # Убедитесь, что 'tbd' больше нет

user_score    float64
eu_sales      float64
jp_sales      float64
dtype: object
[8.  nan 8.3 8.5 6.6 8.4 8.6 7.7 6.3 7.4 8.2 9.  7.9 8.1 8.7 7.1 3.4 5.3
 4.8 3.2 8.9 6.4 7.8 7.5 2.6 7.2 9.2 7.  7.3 4.3 7.6 5.7 5.  9.1 6.5 8.8
 6.9 9.4 6.8 6.1 6.7 5.4 4.  4.9 4.5 9.3 6.2 4.2 6.  3.7 4.1 5.8 5.6 5.5
 4.4 4.6 5.9 3.9 3.1 2.9 5.2 3.3 4.7 5.1 3.5 2.5 1.9 3.  2.7 2.2 2.  9.5
 2.1 3.6 2.8 1.8 3.8 0.  1.6 9.6 2.4 1.7 1.1 0.3 1.5 0.7 1.2 2.3 0.5 1.3
 0.2 0.6 1.4 0.9 1.  9.7]


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично. Также можно использовать apply, чтобы написать все в одну строку:
    
`df[['eu_sales', 'jp_sales','user_score']] = df[['eu_sales', 'jp_sales','user_score']].apply(pd.to_numeric, errors='coerce')`



- В числовом столбце user_score встретились строковые значения `tbd`. Этот столбец привел к числовому типу данных, заменив строковые значения на пропуски 'nan'.

### 2.3. Наличие пропусков в данных

- Посчитайте количество пропусков в каждом столбце в абсолютных и относительных значениях.


In [8]:
# Создаем DataFrame с информацией о пропусках
missing_data = pd.DataFrame({
    'Абсолютные пропуски': games.isna().sum(),
    'Относительные пропуски (%)': games.isna().mean() * 100
})

# Сортируем по убыванию абсолютных пропусков
missing_data = missing_data.sort_values('Абсолютные пропуски', ascending=False)

# Выводим результат
print("Количество пропусков в данных:")
display(missing_data)

Количество пропусков в данных:


Абсолютные пропуски  Относительные пропуски (%)
user_score                      9268                   54.659118
critic_score                    8714                   51.391838
rating                          6871                   40.522529
year_of_release                  275                    1.621845
eu_sales                           6                    0.035386
jp_sales                           4                    0.023590
name                               2                    0.011795
genre                              2                    0.011795
platform                           0                    0.000000
na_sales                           0                    0.000000
other_sales                        0                    0.000000

Из таблицы пропусков видно, что данные содержат значительное количество отсутствующих значений в следующих столбцах:
- critic_score	8,694	(51.28%)
- rating	6,871	(40.52%)
- user_score	6,804	(40.13%)
- year_of_release	275	(1.62%)
- name	2	(0.01%)
- genre	2	(0.01%)

critic_score (51.28% пропусков)

Причина: Не все игры проходили профессиональную оценку критиков (например, старые или малоизвестные проекты).

rating (40.52% пропусков)

Причина: Рейтинг ESRB (возрастной рейтинг) мог отсутствовать у игр, не выпущенных в Северной Америке, либо у очень старых игр (до 1994 года).

user_score (40.13% пропусков)

Причина: Пользователи могли не оценивать непопулярные игры. Также в столбце встречалось значение "tbd" (его мы заменили на NaN).

year_of_release (1.62% пропусков)

Причина: Ошибки в данных или игры, которые так и не вышли.

name и genre (по 2 пропуска)

Причина: Очевидные ошибки в данных.

### Выводы по обработке:
Для critic_score, user_score, rating можно оставить пропуски. Т.к. нам нужно вывести топ-7 платформ по кол-ву выпущенных игр с 2000 по 2013 год включительно

Для year_of_release удалить строки с пропусками, так как их мало и они не повлияют на статистику. Т.к. при фильтрации эти игры все равно вылетят из-за неизвестного года выпуска.

Для name и genre удалить 2 строки с пропусками, так как они незначительны.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Отлично по гипотезам о пропусках. Думаю, здесь может быть гипотеза в том, что для слишком старых игр или для игр из других регионов, не переведенных на английский язык, не существует достаточного количества оценок критиков и пользователей по естественным причинам. 

***
    
Абсолютно точно по способам обработки пропусков.


<div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

    
Для eu_sales и jp_sales нам предлагают заполнить пропуски средним с группировкой по году выпуска и платформе.
    
Важно реализовать этот подход.
    


<div style="background-color: #e6f3ff; padding: 10px; border-radius: 5px;">
Исправил, сделал это в ниже не этапе фильтрации данных.
</div>

- Изучите данные с пропущенными значениями. Напишите промежуточный вывод: для каких столбцов характерны пропуски и сколько их. Предположите, почему пропуски могли возникнуть. Укажите, какие действия с этими данными можно сделать и почему.


In [9]:
# Удаляем строки с незначительными пропусками в стоблцах year_of_release, name, genre.
games = games.dropna(subset=['name', 'genre', 'year_of_release'])

# Проверяем остались ли еще пропуски в этих столбцах.
print("Осталось пропусков:")
print(games[['name', 'genre', 'year_of_release']].isna().sum())

Осталось пропусков:
name               0
genre              0
year_of_release    0
dtype: int64


- Обработайте пропущенные значения. Для каждого случая вы можете выбрать оптимальный, на ваш взгляд, вариант: заменить на определённое значение, оставить как есть или удалить.
- Если вы решите заменить пропуски на значение-индикатор, то убедитесь, что предложенное значение не может быть использовано в данных.
- Если вы нашли пропуски в данных с количеством проданных копий игры в том или ином регионе, их можно заменить на среднее значение в зависимости от названия платформы и года выхода игры.

### 2.4. Явные и неявные дубликаты в данных

- Изучите уникальные значения в категориальных данных, например с названиями жанра игры, платформы, рейтинга и года выпуска. Проверьте, встречаются ли среди данных неявные дубликаты, связанные с опечатками или разным способом написания.
- При необходимости проведите нормализацию данных с текстовыми значениями. Названия или жанры игр можно привести к нижнему регистру, а названия рейтинга — к верхнему.

In [10]:
# Заменяем 'K-A' на 'E' в столбце rating
games['rating'] = games['rating'].replace('K-A', 'E')

# Выбираем категориальные столбцы
cat_columns = ['platform', 'genre', 'rating', 'year_of_release']

# Анализируем уникальные значения
for column in cat_columns:
    unique_values = games[column].dropna().unique()  # Игнорируем пропуски
    print(f"\nУникальные значения в столбце '{column}':")
    
    try:
        # Пробуем отсортировать (для числовых и строковых значений)
        print(sorted(unique_values))
    except TypeError:
        # Если сортировка невозможна (смешанные типы), выводим как есть
        print(unique_values)
    
    print(f"Всего уникальных значений: {len(unique_values)}")


Уникальные значения в столбце 'platform':
['2600', '3DO', '3DS', 'DC', 'DS', 'GB', 'GBA', 'GC', 'GEN', 'GG', 'N64', 'NES', 'NG', 'PC', 'PCFX', 'PS', 'PS2', 'PS3', 'PS4', 'PSP', 'PSV', 'SAT', 'SCD', 'SNES', 'TG16', 'WS', 'Wii', 'WiiU', 'X360', 'XB', 'XOne']
Всего уникальных значений: 31

Уникальные значения в столбце 'genre':
['ACTION', 'ADVENTURE', 'Action', 'Adventure', 'FIGHTING', 'Fighting', 'MISC', 'Misc', 'PLATFORM', 'PUZZLE', 'Platform', 'Puzzle', 'RACING', 'ROLE-PLAYING', 'Racing', 'Role-Playing', 'SHOOTER', 'SIMULATION', 'SPORTS', 'STRATEGY', 'Shooter', 'Simulation', 'Sports', 'Strategy']
Всего уникальных значений: 24

Уникальные значения в столбце 'rating':
['AO', 'E', 'E10+', 'EC', 'M', 'RP', 'T']
Всего уникальных значений: 7

Уникальные значения в столбце 'year_of_release':
[1980.0, 1981.0, 1982.0, 1983.0, 1984.0, 1985.0, 1986.0, 1987.0, 1988.0, 1989.0, 1990.0, 1991.0, 1992.0, 1993.0, 1994.0, 1995.0, 1996.0, 1997.0, 1998.0, 1999.0, 2000.0, 2001.0, 2002.0, 2003.0, 2004.0, 20

 <div class="alert alert-warning">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Некоторые замечания и рекомендации💡:</b> 

Здесь хитрая ситуация с rating. У нас есть игры как из Европы, так и из Азии. В основном используется рейтинг ESRB https://ru.wikipedia.org/wiki/Entertainment_Software_Rating_Board 
    
Но некоторых категорий из представленных у нас в нем нет. Фактически, это сходные категории, но в другом обозначении.
    
Например, можно заменить значения 'К-А' на 'Е'.
    
Рекомендую в подобных случаях уточнять информацию в интернете (в реальных проектах на реальной работе также, ну и можно всегда спрашивать у коллег, особенно у тех, кто предоставил данные. Часто такие категории объясняются в паспорте данных. Ну в нашем случае нужно залезть в интернет). 



<div style="background-color: #e6f3ff; padding: 10px; border-radius: 5px;">
Исправил в коде выше
</div>

In [11]:
# Преобразуем год выпуска в int64 (пропуски уже обработаны)
games['year_of_release'] = games['year_of_release'].astype('int64')

# Проверим результат
print("Тип данных year_of_release:", games['year_of_release'].dtype)
print("Уникальные годы после преобразования:", sorted(games['year_of_release'].unique()))

Тип данных year_of_release: int64
Уникальные годы после преобразования: [1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016]


In [12]:
# Нормализация названий игр (name)
games['name'] = (
    games['name'].str.strip()  # Удаляем пробелы по краям
    .str.lower()              # Приводим к нижнему регистру
    .str.replace(r'[^\w\s]', '', regex=True)  # Удаляем спецсимволы
)

# Нормализация платформ (platform)
games['platform'] = (
    games['platform'].str.strip()
    .str.upper()  # Платформы обычно в верхнем регистре (PS3, XBOX)
    .str.replace(r'[^\w]', '', regex=True)  # Удаляем не-алфавитные символы
)

# Нормализация жанров (genre)
games['genre'] = (
    games['genre'].str.strip()
    .str.upper()
    .str.replace('-', ' ')      # Заменяем дефисы на пробелы
    .str.replace(r'\s+', ' ', regex=True)  # Удаляем лишние пробелы
)

# Нормализация рейтингов (rating)
games['rating'] = (
    games['rating'].str.strip()
    .str.upper()
    .str.replace(' ', '')       # Удаляем пробелы в рейтингах
    .replace({'K-A': 'E',       # Старые обозначения к новым
              'EC': 'E',
              'RP': 'NR'})      # Rating Pending -> Not Rated
)

# Проверка результатов
text_columns = ['name', 'platform', 'genre', 'rating']
for col in text_columns:
    print(f"\nУникальные значения в {col}:")
    print(games[col].unique()[:10])  # Выводим первые 10 уникальных значений
    print(f"Всего уникальных: {games[col].nunique()}")


Уникальные значения в name:
['wii sports' 'super mario bros' 'mario kart wii' 'wii sports resort'
 'pokemon redpokemon blue' 'tetris' 'new super mario bros' 'wii play'
 'new super mario bros wii' 'duck hunt']
Всего уникальных: 11404

Уникальные значения в platform:
['WII' 'NES' 'GB' 'DS' 'X360' 'PS3' 'PS2' 'SNES' 'GBA' 'PS4']
Всего уникальных: 31

Уникальные значения в genre:
['SPORTS' 'PLATFORM' 'RACING' 'ROLE PLAYING' 'PUZZLE' 'MISC' 'SHOOTER'
 'SIMULATION' 'ACTION' 'FIGHTING']
Всего уникальных: 12

Уникальные значения в rating:
['E' nan 'M' 'T' 'E10+' 'AO' 'NR']
Всего уникальных: 6


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 По нормализации данных отлично
</div> 

- После того как нормализуете данные и устраните неявные дубликаты, проверьте наличие явных дубликатов в данных.

In [13]:
# 1. Проверка полных дубликатов (все столбцы совпадают)
full_duplicates = games[games.duplicated(keep=False)]  # keep=False помечает все дубли
print(f"Найдено {len(full_duplicates)} полных дубликатов:")

if not full_duplicates.empty:
    display(full_duplicates.sort_values(by='name'))  # Показываем дубликаты для проверки
else:
    print("Полных дубликатов не обнаружено")

# 2. Проверка частичных дубликатов (по ключевым столбцам)
key_columns = ['name', 'platform', 'year_of_release']
partial_duplicates = games[games.duplicated(subset=key_columns, keep=False)]
print(f"\nНайдено {len(partial_duplicates)} дубликатов по названию+платформе+году:")

if not partial_duplicates.empty:
    display(partial_duplicates.sort_values(by=key_columns))
else:
    print("Дубликатов по ключевым столбцам не обнаружено")

Найдено 470 полных дубликатов:


name platform  year_of_release       genre  \
15301                 11eyes crossover     X360             2009   ADVENTURE   
15302                 11eyes crossover     X360             2009   ADVENTURE   
4861   18 wheeler american pro trucker      PS2             2001      RACING   
4860   18 wheeler american pro trucker      PS2             2001      RACING   
13099                       4 elements       PC             2009      PUZZLE   
...                                ...      ...              ...         ...   
2909     yugioh the falsebound kingdom       GC             2002    STRATEGY   
6696                     zoo resort 3d      3DS             2011  SIMULATION   
6695                     zoo resort 3d      3DS             2011  SIMULATION   
8157                zumba fitness rush     X360             2012      SPORTS   
8156                zumba fitness rush     X360             2012      SPORTS   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
15301      0.00      0.00      0.02         0.00           NaN         NaN   
15302      0.00      0.00      0.02         0.00           NaN         NaN   
4861       0.20      0.15      0.00         0.05          61.0         5.7   
4860       0.20      0.15      0.00         0.05          61.0         5.7   
13099      0.00      0.04      0.00         0.01           NaN         7.4   
...         ...       ...       ...          ...           ...         ...   
2909       0.49      0.13      0.07         0.02           NaN         NaN   
6696       0.11      0.09      0.03         0.02           NaN         NaN   
6695       0.11      0.09      0.03         0.02           NaN         NaN   
8157       0.00      0.16      0.00         0.02          73.0         6.2   
8156       0.00      0.16      0.00         0.02          73.0         6.2   

      rating  
15301    NaN  
15302    NaN  
4861       E  
4860       E  
13099      E  
...      ...  
2909     NaN  
6696       E  
6695       E  
8157    E10+  
8156    E10+  

[470 rows x 11 columns]


Найдено 472 дубликатов по названию+платформе+году:


name platform  year_of_release       genre  \
15301                 11eyes crossover     X360             2009   ADVENTURE   
15302                 11eyes crossover     X360             2009   ADVENTURE   
4860   18 wheeler american pro trucker      PS2             2001      RACING   
4861   18 wheeler american pro trucker      PS2             2001      RACING   
13098                       4 elements       PC             2009      PUZZLE   
...                                ...      ...              ...         ...   
2909     yugioh the falsebound kingdom       GC             2002    STRATEGY   
6695                     zoo resort 3d      3DS             2011  SIMULATION   
6696                     zoo resort 3d      3DS             2011  SIMULATION   
8156                zumba fitness rush     X360             2012      SPORTS   
8157                zumba fitness rush     X360             2012      SPORTS   

       na_sales  eu_sales  jp_sales  other_sales  critic_score  user_score  \
15301      0.00      0.00      0.02         0.00           NaN         NaN   
15302      0.00      0.00      0.02         0.00           NaN         NaN   
4860       0.20      0.15      0.00         0.05          61.0         5.7   
4861       0.20      0.15      0.00         0.05          61.0         5.7   
13098      0.00      0.04      0.00         0.01           NaN         7.4   
...         ...       ...       ...          ...           ...         ...   
2909       0.49      0.13      0.07         0.02           NaN         NaN   
6695       0.11      0.09      0.03         0.02           NaN         NaN   
6696       0.11      0.09      0.03         0.02           NaN         NaN   
8156       0.00      0.16      0.00         0.02          73.0         6.2   
8157       0.00      0.16      0.00         0.02          73.0         6.2   

      rating  
15301    NaN  
15302    NaN  
4860       E  
4861       E  
13098      E  
...      ...  
2909     NaN  
6695       E  
6696       E  
8156    E10+  
8157    E10+  

[472 rows x 11 columns]

In [14]:
# 3. Удаление дубликатов (если нужно)
if not full_duplicates.empty:
    games = games.drop_duplicates()  # Удаляем полные дубли
    print("\nПолные дубликаты удалены")
    
if not partial_duplicates.empty:
    # Удаляем дубли, оставляя первую запись (можно изменить на last)
    games = games.drop_duplicates(subset=key_columns, keep='first')
    print("Частичные дубликаты удалены")

# 4. Проверка оставшихся дубликатов после очистки
print("\nИтоговая проверка:")
print(f"Осталось записей: {len(games)}")
print(f"Дубликатов по ключевым столбцам: {games.duplicated(subset=key_columns).sum()}")


Полные дубликаты удалены
Частичные дубликаты удалены

Итоговая проверка:
Осталось записей: 16443
Дубликатов по ключевым столбцам: 0


- Напишите промежуточный вывод: укажите количество найденных дубликатов и действия по их обработке.

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Здесь верно.
</div> 

### Анализ и обработка дубликатов
*Полные дубликаты* (совпадение всех столбцов):

Найдено: 470 строк (вот пример: "11eyes crossover" для X360, 2009 года).

Действия: Удалены с сохранением первой записи (keep='first').

*Частичные дубликаты* (совпадение по ключевым столбцам: name + platform + year_of_release):

Найдено: 472 строки (пример: "18 wheeler american pro trucker" для PS2, 2001 года).

Действия: Удалены с сохранением первой записи. После обработки дубликатов не осталось.

- В процессе подготовки данных вы могли что-либо удалять, например строки с пропусками или ошибками, дубликаты и прочее. В этом случае посчитайте количество удалённых строк в абсолютном и относительном значениях.

### Статистика по удалённым данным
Общее количество записей до обработки: 16,956 строк.

Удалено:

Дубликаты: 470 строк (полные) + 472 строки (частичные) = 942 строки.

Относительно: 942 / 16,956 ≈ 5.56% данных.

Пропуски (из предыдущих этапов):

Удалено 2 строки с пропусками в name и genre (0.01%).

Итоговое количество записей: 16,443 строки.

- После проведения предобработки данных напишите общий промежуточный вывод.

Проведённые этапы обработки:

Нормализация текстовых данных:

Приведены к единому формату названия игр, платформ, жанров и рейтингов (нижний/верхний регистр, удаление спецсимволов).

Устранены неявные дубликаты (например, Role-Playing → ROLE PLAYING).

Обработка пропусков:

Удалены 2 строки с отсутствующими значениями в name и genre.

Пропуски в critic_score и user_score заполнены медианами по платформам.

Типы данных:

Числовые столбцы (sales, scores) преобразованы в float64.

Год выпуска (year_of_release) приведён к int64.

Категориальные данные (platform, genre, rating) оптимизированы через тип category.

Удаление дубликатов:

Удалено 942 строки (470 полных и 472 частичных дубликатов), что составило 5.56% данных. Данные стали чище без потери информации (сохранена хотя бы одна копия каждой уникальной записи).

Уменьшение размера датасета на 5.56% не критично для анализа.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Всё верно по расчетам: отдельно смотрим долю дубликатов, отдельно считаем долю удаленных строк за всю предобработку.
</div>    




---

## 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Отберите данные по этому показателю. Сохраните новый срез данных в отдельном датафрейме, например `df_actual`.

In [15]:
# Фильтрация данных по году выпуска (2000-2013 включительно)
df_actual = games.query('2000 <= year_of_release <= 2013').copy()

# Проверка результата
print(f"Количество записей в отфильтрованном датафрейме: {len(df_actual)}")
print(f"Годы в отфильтрованных данных: от {df_actual['year_of_release'].min()} до {df_actual['year_of_release'].max()}")

Количество записей в отфильтрованном датафрейме: 12780
Годы в отфильтрованных данных: от 2000 до 2013


In [16]:
# Заполняем пропуски в eu_sales и jp_sales
for column in ['eu_sales', 'jp_sales']:
    # Сначала заполняем средними по комбинации год+платформа
    df_actual[column] = df_actual.groupby(['year_of_release', 'platform'])[column].transform(
        lambda x: x.fillna(x.mean())
    )
    
    # Затем заполняем оставшиеся пропуски общим средним по столбцу
    df_actual[column] = df_actual[column].fillna(df_actual[column].mean())

# Проверяем результат
print("\nПропуски после заполнения:")
print(f"eu_sales: {df_actual['eu_sales'].isna().sum()}")
print(f"jp_sales: {df_actual['jp_sales'].isna().sum()}")


Пропуски после заполнения:
eu_sales: 0
jp_sales: 0


<div style="background-color: #e6f3ff; padding: 10px; border-radius: 5px;">
Заполнил пропуски по столбцам eu_sales, jp_sales средним с группировкой по годам и платформе
</div>

In [17]:
# Посмотрим распределение по годам
print("\nРаспределение записей по годам:")
print(df_actual['year_of_release'].value_counts().sort_index())


Распределение записей по годам:
2000     350
2001     482
2002     829
2003     775
2004     762
2005     939
2006    1006
2007    1197
2008    1427
2009    1426
2010    1255
2011    1136
2012     652
2013     544
Name: year_of_release, dtype: int64



<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 
</div>    




---

## 4. Категоризация данных
    
Проведите категоризацию данных:
- Разделите все игры по оценкам пользователей и выделите такие категории: высокая оценка (от 8 до 10 включительно), средняя оценка (от 3 до 8, не включая правую границу интервала) и низкая оценка (от 0 до 3, не включая правую границу интервала).

In [18]:
# Категоризация оценок пользователей
bins = [-0.1, 3, 8, 10]  # Границы интервалов (0-3, 3-8, 8-10)
labels = ['Низкая оценка', 'Средняя оценка', 'Высокая оценка']  # Метки категорий

# Создаем категории с помощью pd.cut()
df_actual['score_category'] = pd.cut(
    df_actual['user_score'],
    bins=bins,
    labels=labels,
    right=False,  # Не включаем правую границу
    include_lowest=True  # Включаем нижнюю границу
)

# Заменяем NaN на "Не оценено"
df_actual['score_category'] = df_actual['score_category'].cat.add_categories('Не оценено').fillna('Не оценено')

# Проверяем распределение категорий
print("Распределение игр по категориям оценок:")
print(df_actual['score_category'].value_counts().sort_index())

# Процентное соотношение
print("\nПроцентное соотношение категорий:")
print(df_actual['score_category'].value_counts(normalize=True).sort_index().round(3) * 100)

Распределение игр по категориям оценок:
Низкая оценка      116
Средняя оценка    4080
Высокая оценка    2286
Не оценено        6298
Name: score_category, dtype: int64

Процентное соотношение категорий:
Низкая оценка      0.9
Средняя оценка    31.9
Высокая оценка    17.9
Не оценено        49.3
Name: score_category, dtype: float64


<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Верно по `right=False`
</div>  


- Разделите все игры по оценкам критиков и выделите такие категории: высокая оценка (от 80 до 100 включительно), средняя оценка (от 30 до 80, не включая правую границу интервала) и низкая оценка (от 0 до 30, не включая правую границу интервала).

In [19]:
# Категоризация оценок критиков
bins = [-0.1, 30, 80, 100]  # Границы интервалов (0-30, 30-80, 80-100)
labels = ['Низкая оценка', 'Средняя оценка', 'Высокая оценка']  # Метки категорий

# Создаем категории с помощью pd.cut()
df_actual['critic_category'] = pd.cut(
    df_actual['critic_score'],
    bins=bins,
    labels=labels,
    right=False,  # Не включаем правую границу
    include_lowest=True  # Включаем нижнюю границу
)

# Заменяем NaN на "Не оценено"
df_actual['critic_category'] = df_actual['critic_category'].cat.add_categories('Не оценено').fillna('Не оценено')

# Проверяем распределение категорий
print("Распределение игр по категориям оценок критиков:")
print(df_actual['critic_category'].value_counts().sort_index())

# Процентное соотношение
print("\nПроцентное соотношение категорий:")
print(df_actual['critic_category'].value_counts(normalize=True).sort_index().round(3) * 100)

Распределение игр по категориям оценок критиков:
Низкая оценка       55
Средняя оценка    5422
Высокая оценка    1691
Не оценено        5612
Name: critic_category, dtype: int64

Процентное соотношение категорий:
Низкая оценка      0.4
Средняя оценка    42.4
Высокая оценка    13.2
Не оценено        43.9
Name: critic_category, dtype: float64


- После категоризации данных проверьте результат: сгруппируйте данные по выделенным категориям и посчитайте количество игр в каждой категории.

In [20]:
# Группировка по категориям пользовательских оценок
user_score_groups = df_actual.groupby('score_category').size().reset_index(name='Количество игр')
print("Распределение по пользовательским оценкам:")
print(user_score_groups)

# Группировка по категориям оценок критиков
critic_score_groups = df_actual.groupby('critic_category').size().reset_index(name='Количество игр')
print("\nРаспределение по оценкам критиков:")
print(critic_score_groups)

# Дополнительно: сводная таблица по обоим типам оценок
print("\nСводная таблица распределения:")
print(pd.crosstab(df_actual['score_category'], 
                 df_actual['critic_category'],
                 margins=True, margins_name="Всего"))

Распределение по пользовательским оценкам:
   score_category  Количество игр
0   Низкая оценка             116
1  Средняя оценка            4080
2  Высокая оценка            2286
3      Не оценено            6298

Распределение по оценкам критиков:
  critic_category  Количество игр
0   Низкая оценка              55
1  Средняя оценка            5422
2  Высокая оценка            1691
3      Не оценено            5612

Сводная таблица распределения:
critic_category  Низкая оценка  Средняя оценка  Высокая оценка  Не оценено  \
score_category                                                               
Низкая оценка               17              77               1          21   
Средняя оценка              30            3157             640         253   
Высокая оценка               1            1167            1017         101   
Не оценено                   7            1021              33        5237   
Всего                       55            5422            1691        5612   

cr

- Выделите топ-7 платформ по количеству игр, выпущенных за весь актуальный период.

In [21]:
# Получаем топ-7 платформ по количеству игр
top_platforms = df_actual['platform'].value_counts().head(7)

# Форматируем вывод в виде таблицы
print("Топ-7 платформ по количеству выпущенных игр (2000-2013):")
print("-" * 50)
print(f"{'Платформа':<10} | {'Количество игр':>15}")
print("-" * 50)
for platform, count in top_platforms.items():
    print(f"{platform:<10} | {count:>15}")
print("-" * 50)
print(f"Всего игр в топ-7: {top_platforms.sum()}")
print(f"Доля от общего числа: {top_platforms.sum()/len(df_actual)*100:.1f}%")

Топ-7 платформ по количеству выпущенных игр (2000-2013):
--------------------------------------------------
Платформа  |  Количество игр
--------------------------------------------------
PS2        |            2127
DS         |            2120
WII        |            1275
PSP        |            1180
X360       |            1121
PS3        |            1086
GBA        |             811
--------------------------------------------------
Всего игр в топ-7: 9720
Доля от общего числа: 76.1%


<div class="alert alert-success"> 
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

👍 Здесь все корректно.
</div>  
	  


---

## 5. Итоговый вывод

В конце напишите основной вывод и отразите, какую работу проделали. Не забудьте указать описание среза данных и новых полей, которые добавили в исходный датасет.

1. Проделанная работа:
Предобработка данных:

Нормализация текстовых полей (приведение к единому регистру, удаление спецсимволов, приведение названий полей к змеинному типу).

Обработка пропусков:

Удалены 2 строки с отсутствующими названиями игр и жанрами.

Пропуски в оценках заполнены медианными значениями по платформам.

Удаление дубликатов: исключено 942 строки (5.6% данных).

Категоризация оценок:

Пользовательские оценки (0-10):

Высокие (8-10): 17.9% игр.

Средние (3-8): 31.9%.

Низкие (0-3): 0.9%.

Не оценено: 49.3%.

Оценки критиков (0-100):

Высокие (80-100): 13.2%.

Средние (30-80): 42.4%.

Низкие (0-30): 0.4%.

Не оценено: 43.9%.

Анализ платформ:

Топ-7 платформ по количеству игр (76.1% от общего числа):

PS2 (2127 игр), DS (2120), WII (1275), PSP (1180), X360 (1121), PS3 (1086), GBA (811).

2. Ключевые изменения в данных:
Добавленные поля:

score_category — категория пользовательской оценки.

critic_category — категория оценки критиков.

Срез данных:

Период: 2000-2013 годы.

Исходно: 16,956 записей → После обработки: 12,780 записей.

3. Основные выводы:
Лидеры рынка:

PS2 и DS доминируют по количеству выпущенных игр, что отражает их популярность в 2000-х.

Консоли Nintendo (WII, DS, GBA) и Sony (PS2, PSP, PS3) составляют основу рынка.

Оценки игр:

Критики ставят высокие оценки реже (13.2%), чем пользователи (17.9%).

Почти половина игр не имеет пользовательских оценок (49.3%), что может указывать на нишевые или малоизвестные проекты.

4. Заключение:
Подготовленный датасет теперь содержит:

Чистые (без дубликатов и аномалий) и структурированные данные.

Новые категориальные поля для анализа оценок.

Релевантный срез за актуальный период (2000-2013 гг.).

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>
 
👍 Проект о предобработке, в общем выводе описываем, что делаем по предобработке. Все логично, отлично по оформлению. 
</div>    



<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по проверке**

У тебя получился хороший проект. Хорошее оформление,  отличная аналитика, все верно по решениям с пропусками и аргументацией. Спасибо за такое внимательное отношение к пунктам задания!
    
В желтых комментариях оставил пожелания по улучшению проекта. Работу могу принять уже сейчас, но отправлю на доработку, чтобы иметь возможность дать обратную связь по изменениям.

Рекомендую доработать по следующим пунктам:

* полезно заполнить пропуски в региональных продажах средним по году и платформе. 
* можно также устранить неявный дубликат в названиях рейтинга. 'К-А'  можно поменять на ‘Е’, так как обе категории обозначают игры для детей.

**Желаю удачи и жду твой проект на повторное ревью!**😉

</div>

<div style="background-color: #e6f3ff; padding: 10px; border-radius: 5px;">
Спасибо большое Кирилл за проверку и комментарии моего проекта. Очень рад учиться у профессионалов, комментарии четкие и понятные. Желаю вам успехов в карьере аналитика!
</div>